examples/automm/memory_bank/utils.py

```diff
@@ -0,0 +1,199 @@
+import torch
+import pandas as pd
+from tqdm import tqdm
+import torch.nn.functional as F
+import torch.nn as nn
+from autogluon.multimodal import MultiModalPredictor
+from datasets import list_datasets, load_dataset
+from setfit import sample_dataset
+from imagedatasets import build_dataset
+from torch.utils.data import Dataset as TorchDataset
+
+
+def cls_acc(output, target, topk=1):
+    pred = output.topk(topk, 1, True, True)[1].t()
+    correct = pred.eq(target.view(1, -1).expand_as(pred))
+    acc = float(correct[: topk].reshape(-1).float().sum(0, keepdim=True).cpu().numpy())
+    acc = 100 * acc / target.shape[0]
+    return acc
+
+def generate_image_df(args, dataset):
+    column_names = args.column_names + [args.label_column]
+    dataset_df = pd.DataFrame(columns=column_names)
+    for data in dataset:
+        per_df = pd.DataFrame([[data.impath, data.label]], columns=column_names)
+        dataset_df = pd.concat(
+            [dataset_df, per_df],
+            ignore_index=True,
+        )
+    return dataset_df
+    
+def generate_dataset(args):
+    if args.type == "text":
+        dataset = load_dataset(args.dataset)
+        train_dataset = sample_dataset(dataset["train"], label_column=args.label_column, num_samples=args.shots)
+        val_dataset = dataset["validation"] if "validaiton" in dataset else dataset["test"]
+        test_dataset = dataset["test"]
+
+        train_dataset.set_format("pandas")
+        val_dataset.set_format("pandas")
+        test_dataset.set_format("pandas")
+
+        train_df = train_dataset[:].drop(columns=["label_text"])
+        val_df = val_dataset[:].drop(columns=["label_text"])
+        test_df = test_dataset[:].drop(columns=["label_text"])
+
+        num_classes = train_dataset[args.label_column].max() + 1
+    else:
+        dataset = build_dataset(args.dataset, args.data_path, args.shots)
+        train_df = generate_image_df(args, dataset.train_x)
+        val_df = generate_image_df(args, dataset.val)
+        test_df = generate_image_df(args, dataset.test)
+        num_classes = len(dataset.classnames)
+    
+    return dataset, train_df, val_df, test_df, num_classes
+
+
+def extract_embedding(args, data, predictor, column_names):
+    features = []
+    for per_name in column_names:
+        per_features = torch.tensor(predictor.extract_embedding({per_name: data[per_name]})[per_name]).cuda()
+        features.append(per_features)
+    features = torch.stack(features, dim=1).cuda()
+    features = features.mean(dim=1)
+    features /= features.norm(dim=-1, keepdim=True)
+    return features
+
+
+def generate_clip_weights(args, classnames, template, predictor):
```

Comment:
We can consider to just introduce the variable called `semantic_label_embedding` and compare the similarity between `semantic_label_embedding` and the embeddings of the labels stored in the memory bank.